src/pudl/io_managers.py

```diff
@@ -341,6 +341,69 @@ def handle_output(self, context: OutputContext, obj: pd.DataFrame | str):
                 "queries."
             )
 
+    def load_input(self, context: InputContext) -> pd.DataFrame:
+        """Load a dataframe from a sqlite database.
+
+        Args:
+            context: dagster keyword that provides access output information like asset
+                name.
+        """
+        table_name = self._get_table_name(context)
+        _ = self._get_sqlalchemy_table(table_name)
+
+        engine = self.engine
+
+        with engine.connect() as con:
+            try:
+                df = pd.read_sql_table(table_name, con)
+            except ValueError:
+                raise ValueError(
+                    f"{table_name} not found. Either the table was dropped "
+                    "or it doesn't exist in the pudl.metadata.resources."
+                    "Add the table to the metadata and recreate the database."
+                )
+            if df.empty:
+                raise AssertionError(
+                    f"The {table_name} table is empty. Materialize "
+                    "the {table_name} asset so it is available in the database."
+                )
+            return df
+
+
+class PudlSQLiteIOManager(SQLiteIOManager):
+    """IO Manager that writes and retrieves dataframes from a SQLite database.
+
+    This class extends the SQLiteIOManager class to manage database metadata and dtypes
+    using the :func:`pudl.metadata.classes.Package` class.
+    """
+
+    def __init__(
+        self,
+        base_dir: str = None,
+        db_name: str = None,
+        package: Package = None,
+        timeout: float = 1_000.0,
+    ):
+        """Initialize PudlSQLiteIOManager.
+
+        Args:
+            base_dir: base directory where all the step outputs which use this object
+                manager will be stored in.
+            db_name: the name of sqlite database.
+            package: Tabular datapackage used to create sqlalchemy metadata and check
+                datatypes. If not specified, defaults to metadata stored in the
+                pudl.metadata subpackage.
+            timeout: How many seconds the connection should wait before raising an
+                exception, if the database is locked by another connection.  If another
+                connection opens a transaction to modify the database, it will be locked
+                until that transaction is committed.
+        """
+        if not package:
+            package = Package.from_resource_ids()
+        self.package = package
+        md = self.package.to_sql()
+        super().__init__(base_dir, db_name, md, timeout)
+
     def load_input(self, context: InputContext) -> pd.DataFrame:
```

Comment:
In #2459 @zaneselvans you'll probably rewrite the `PudlSQLiteIOManager.load_input()` method to:
1. Make sure the `table_name` exists as a resource in `self.package`.
2. Apply `enforce_schema()` to the df instead of `apply_pudl_dtypes()`.

You'll also need to overwrite the `_handle_pandas_output()` method to check the table exists in self.md and enforce schema.

Reply:
In `PudlSQLiteIOManager._handle_pandas_output()` does it make sense to use `enforce_schema()`?

If it calls the parent `_handle_pandas_output()` then `apply_pudl_dtypes()` will be called on the dataframe, and I think that will remove the categorical type information, and also any data source specific typing information (since the generic call to `apply_pudl_dtypes()` without `group="eia"` or something similar just uses the generic field definitions.  I guess maybe that means we can't call `super()._handle_pandas_output()` and we have to override the whole method, and should use `enforce_schema()` to get the data source specific types, even if the categorical information is lost when we write to SQLite.